
# 多线程处理
很多时候，我们需要在某个接口中，调用其他服务的接口。比如有这样的业务场景：在用户信息查询接口中需要返回：用户名称、性别、等级、头像、积分、成长值等信息。而用户名称、性别、等级、头像在用户服务中，积分在积分服务中，成长值在成长值服务中。为了汇总这些数据统一返回，需要另外提供一个对外接口服务。

于是，用户信息查询接口需要调用用户查询接口、积分查询接口 和 成长值查询接口，然后汇总数据统一返回。

java8以后通过CompleteFuture类实现该功能。我们这里以CompleteFuture为例：

```java
public UserInfo getUserInfo(Long id) throws InterruptedException, ExecutionException {
    final UserInfo userInfo = new UserInfo();
    CompletableFuture userFuture = CompletableFuture.supplyAsync(() -> {
        getRemoteUserAndFill(id, userInfo);
        return Boolean.TRUE;
    }, executor);

    CompletableFuture bonusFuture = CompletableFuture.supplyAsync(() -> {
        getRemoteBonusAndFill(id, userInfo);
        return Boolean.TRUE;
    }, executor);

    CompletableFuture growthFuture = CompletableFuture.supplyAsync(() -> {
        getRemoteGrowthAndFill(id, userInfo);
        return Boolean.TRUE;
    }, executor);
    CompletableFuture.allOf(userFuture, bonusFuture, growthFuture).join();

    userFuture.get();
    bonusFuture.get();
    growthFuture.get();

    return userInfo;
}
```
温馨提醒一下，这两种方式别忘了使用线程池。示例中我用到了executor，表示自定义的线程池，为了防止高并发场景下，出现线程过多的问题。

# 异常处理
## 使用全局异常处理器
有些小伙伴，经常喜欢在 Controller、Service 代码中捕获异常。不管是普通异常 Exception，还是运行时异常 RuntimeException，都使用`try/catch`把它们捕获。
反例：

```java
try {
    checkParam(param);
} catch (BusinessException e) {
    return ApiResultUtil.error(1,"参数错误");
}
```

显然这种做法会造成大量重复的代码。我们在 Controller、Service 等业务代码中，尽可能少捕获异常。这种业务异常处理，应该交给拦截器统一处理。有了这个全局的异常处理器，之前我们在 Controller 或者 Service 中的`try/catch`代码可以去掉。

如果在接口中出现异常，全局的异常处理器会帮我们封装结果，返回给用户。
## 优先使用标准异常
在Java中已经定义了许多比较常用的标准异常，比如下面这张图中列出的这些异常。

- IllegalArgumentException 入参不合法
- IllegalStateException 入参的状态不合法
- UnsupportedOperationException 不支持的操作
- SecurityException 安全异常 返回 401/403
- NullPointerException 空指针异常 返回 500

反例：
```java
public void checkValue(int value) {
    if (value < 0) {
        throw new MyIllegalArgumentException("值不能为负");
    }
}
```
自定义了一个异常表示参数错误。其实，我们可以直接复用已有的标准异常。

正例：
```java
public void checkValue(int value) {
    if (value < 0) {
        throw new IllegalArgumentException("值不能为负");
    }
}
```
## 尽可能捕获具体异常

在你的业务逻辑方法中，有可能需要去处理多种不同的异常。 你可能你会觉得比较麻烦，而直接捕获`Exception`。

反例：

```java
try {
    doSomething();
} catch(Exception e) {
    log.error("doSomething处理失败，原因：",e);
}
```

这样捕获异常太笼统了。 其实doSomething方法中，会抛出`FileNotFoundException`和`IOException`。 这种情况我们最好捕获具体的异常，然后分别做处理。

正例：
 
```java
try {
   doSomething();
} catch(FileNotFoundException e) {
  log.error("doSomething处理失败，文件找不到，原因：",e);
} catch(IOException e) {
  log.error("doSomething处理失败，IO出现了异常，原因：",e);
}
```
这样如果后面出现了上面的异常，我们就非常方便知道是什么原因了。